src/psyclone/psyir/transformations/arrayrange2loop_trans.py

```diff
@@ -0,0 +1,369 @@
+        node must be an assignment. The rightmost range node in each array
+        within the assignment is replaced with a loop index and the
+        assignment is placed within a loop iterating over that
+        index. The bounds of the loop are determined from the bounds
+        of the array range on the left hand side of the assignment.
+
+        :param node: an Assignment node.
+        :type node: :py:class:`psyclone.psyir.nodes.Assignment`
+
+        '''
+        self.validate(node)
+
+        parent = node.parent
+        symbol_table = node.scope.symbol_table
+        loop_variable_name = symbol_table.new_symbol_name(root_name="idx")
+        loop_variable_symbol = DataSymbol(loop_variable_name, INTEGER_TYPE)
+        symbol_table.add(loop_variable_symbol)
+
+        # Replace the rightmost range found in all arrays with the
+        # iterator and use the range from the LHS range for the loop
+        # iteration space.
+        for array in node.walk(Array):
+            for idx, child in reversed(list(enumerate(array.children))):
+                if isinstance(child, Range):
+                    if array is node.lhs:
+                        # Save this range to determine indexing
+                        lhs_range = child
+                    array.children[idx] = Reference(
+                        loop_variable_symbol, parent=array)
+                    break
+        position = node.position
+        # Issue #806: If Loop bounds were a Range we would just
+        # need to provide the range node which would be simpler.
+        loop = Loop.create(loop_variable_symbol, lhs_range.children[0],
+                           lhs_range.children[1], lhs_range.children[2],
+                           [node])
+        parent.children[position] = loop
+        loop.parent = parent
+
+    def __str__(self):
+        return "Convert a PSyIR Array Range to a PSyIR Loop."
+
+    @property
+    def name(self):
+        '''
+        :returns: the name of the transformation as a string.
+        :rtype: str
+
+        '''
+        return type(self).__name__
+
+    def validate(self, node, options=None):
+        '''Perform various checks to ensure that it is valid to apply the
+        ArrayRange2LoopTrans transformation to the supplied PSyIR Node.
+
+        :param node: the node that is being checked.
+        :type node: :py:class:`psyclone.psyir.nodes.Assignment`
+
+        :raises TransformationError: if the node argument is not an \
+            Assignment.
+        :raises TransformationError: if the node argument is an \
+            Assignment whose left had side is not an Array.
+        :raises TransformationError: if the node argument is an \
+            Assignment whose left hand side is an Array that does not \
+            have Range specifying the access to at least one of its \
+            dimensions.
+        :raises TransformationError: if two or more of the loop ranges \
```

Comment:
Oops, fixed indentation.

We do need to check here. Even if the Fortran is correct there are currently cases that we can't deal with and we need to make sure we don't incorrectly transform these. For example: `a(:) = b(::2)`, or `a(:) = b(:)` where `a` starts at index 1 and `b` starts at a different index. Once the above cases are supported (issue #814) then I think I agree, although 1) we should also ensure that any PSyIR that is created is also valid, which we don't currently do and 2) if there were an error in the original code, it would be picked up at run-time, but after transformation we would no longer get a run-time error.

Reply:
Oh yes, that's true.